src/test/java/com/openshift/internal/restclient/DefaultClientFilterIntegrationTest.java

```diff
@@ -0,0 +1,155 @@
+		project = (IProject) client.create(projectRequest);
+
+		buildConfigs.add(createBuildConfigWithLabels(project, "build1", new HashMap<String, String>() {{
+			put("foo", "yes");
+			put("bar", "no");
+			put("baz", "no");
+		}}));
+
+		buildConfigs.add(createBuildConfigWithLabels(project, "build2", new HashMap<String, String>() {{
+			put("foo", "no");
+			put("bar", "yes");
+
+		}}));
+
+		buildConfigs.add(createBuildConfigWithLabels(project, "build3", new HashMap<String, String>() {{
+			put("foo", "yes");
+			put("bar", "yes");
+		}}));
+
+		buildConfigs.add(createBuildConfigWithLabels(project, "build4", new HashMap<>()));
+
+	}
+
+	@AfterClass
+	public void cleanup() {
+		buildConfigs.forEach(bc -> cleanUpResource(client, bc));
+		cleanUpResource(client, project);
+	}
+
+	@Test
+	public void testFilteringWithOneLabel() {
+		List<IBuildConfig> list = client.list(BUILD_CONFIG, project.getNamespace(), new HashMap<String, String>() {{
+					put("foo", "yes");
+				}});
+
+		assertEquals(2, list.size());
+		Set<String> names = list.stream().map(IResource::getName).collect(Collectors.toSet());
+		assertTrue("Should contain build1", names.contains("build1"));
+		assertTrue("Should contain build3", names.contains("build3"));
+
+	}
+
+	@Test
+	public void testFilteringWithTwoLabel() {
+		List<IBuildConfig> list = client.list(BUILD_CONFIG, project.getNamespace(), new HashMap<String, String>() {{
+					put("foo", "yes");
+					put("bar", "no");
+				}});
+
+		assertEquals(1, list.size());
+		IBuildConfig bc = list.get(0);
+		assertEquals("build1", bc.getName());
+	}
+
+	@Test
+	public void testFilteringWithLabelExist() {
+		List<IBuildConfig> list = client.list(BUILD_CONFIG, project.getNamespace(), "baz");
+
+		assertEquals(1, list.size());
+		IBuildConfig bc = list.get(0);
+		assertEquals("build1", bc.getName());
+	}
+
+	@Test
+	public void testFilteringWithLabelNotExist() {
+		List<IBuildConfig> list =
+				client.list(BUILD_CONFIG, project.getNamespace(), "!baz");
+
+		assertEquals(2, list.size());
+		Set<String> names = list.stream().map(IResource::getName).collect(Collectors.toSet());
+		assertTrue("Should contain build2", names.contains("build2"));
+		assertTrue("Should contain build3", names.contains("build3"));
+
+	}
+
+	@Test
+	public void testFilteringWithLabelNotEqualTo() {
+		List<IBuildConfig> list = client.list(BUILD_CONFIG, project.getNamespace(), "foo!=yes");
+
+		assertEquals(1, list.size());
+		IBuildConfig bc = list.get(0);
+		assertEquals("build2", bc.getName());
+	}
+
+	@Test
+	public void testFilteringWithLabelCombinedLabelQuery() {
+		List<IBuildConfig> list = client.list(BUILD_CONFIG, project.getNamespace(), "foo,bar=no");
+
+		assertEquals(1, list.size());
+		IBuildConfig bc = list.get(0);
+		assertEquals("build1", bc.getName());
+	}
+
+
+
+	private BuildConfig createBuildConfigWithLabels(IProject project, String name, HashMap<String, String> labelFilter) {
+		BuildConfig bc = factory.stub(BUILD_CONFIG, name);
```

Comment:
Consider replacing with the the capability builder as described here: https://github.com/openshift/openshift-restclient-java